### DISCUSSION TITLE
How to put different header on each page?

### DISCUSSION THREAD

**Discussion by awesome1128** [dompdf/dompdf#3178] [Ideas]
How are you?
I am trying to build the pdf using this library. Works well!!!
But I have an issue with a header section.
Actually, I want to put a different header on each page.
Is this possible?
Please let me know, thank you! 💪

**Comment by bsweeney** [comment]
What do you mean by a different header on each page? Like, every page will have different header content?

The typical way to style a header is to use an element styled `position: fixed;`. As an example:

```
<div style="position: fixed; left: 0px; top: -50px; height: 50px; background-color: orange;">header</div>
<p>some page content</p>
```

But if you want each page to use different content then, and you know where the page starts, you can add content with `position: absolute;`. For example:

```
<div style="position: absolute; left: 0px; top: -50px; height: 50px; background-color: orange;">header 1</div>
<p>some page content</p>
<div style="page-break-before: always;"></div>
<div style="position: absolute; left: 0px; top: -50px; height: 50px; background-color: orange;">header 2</div>
<p>some page content</p>
```

Without a sample of what you want it's hard to provide much more guidance.

**Reply by awesome1128** [reply]
Hi @bsweeney 
Thank you for your kind reply.
Yes, every page will have different header content.

Please see this screenshot.
https://prnt.sc/3pFyEgbjMQL1

I have a design like this.
Let's say, for example, I have 2 pages, so I want to put different header content on every page, but for example, first page may have a lot of content, so it might overflow.

So what I want to know is...
1. how to have different header content on every page. Page 1 has unique content and page 2 has unique content.
2. What if a first or second page has a lot of content, then how can I handle the header content? For example, if first page content overflows, then first page header should be repeated.

I hope you understand what I explained.
Thank you so much!

**Reply by bsweeney** [reply]
Dompdf as part of the normal rendering process is unable to accommodate this type of document. You have two options:

1. Render each section as a separate document then combine each individual PDF into a single one later.
2. Use callbacks to make the appropriate header visible for each section.

The second option, while possible, is a bit more challenging, at least with logic I can think of right now. The document must be structured in a particular way. Here's an example using Dompdf 2.0.3.

HTML:
```
<style>
    .header {
        position: fixed;
        top: -.5in;
        left: 0in;
        right: 0in;
        z-index: -1
    }
    #header1 {
        background-color: #66000033;
    }
    #header2 {
        background-color: #00006633;
    }

    .section {
        page-break-after: always;
    }
    .section:last-of-type() {
        page-break-after: auto;
    }
</style>

<div class="header" id="header1" data-header="1">
    Section 1 Header
</div>
<div class="section" id="section1">
    <h1>Section 1</h1>
    <p>lots and lots of text</p>
    <div style="page-break-before: always;"></div>
    <p>more text</p>
    <div class="end-of-section"></div>
</div>

<div class="header" id="header2" data-header="2">
    Section 2 Header
</div>
<div class="section" id="section2">
    <h1>Section 2</h1>
    <div class="end-of-section"></div>
</div>
```

PHP:
```
$dompdf = new Dompdf();
$section = 1;
$dompdf->setCallbacks([
    [
        'event' => 'begin_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$section) {
            $node = $frame->get_node();
            $header = $node->hasAttributes() && ($data = $node->attributes->getNamedItem("data-header")) !== null ? $data->nodeValue : null;
            if (isset($header)) {
                $style = $frame->get_style();
                if ($header == $section) {
                    $style->set_used("display", "block");
                } else {
                    $style->set_used("display", "none");
                }
            }
        }
    ],
    [
        'event' => 'end_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$section) {
            $node = $frame->get_node();
            $class = $node->hasAttributes() && ($class = $node->attributes->getNamedItem("class")) !== null ? $class->nodeValue : "";
            if ($class === "end-of-section") {
                $section++;
            }
        }
    ]
]);
$dompdf->render();
$dompdf->stream();
```

**Comment by procodix** [comment]
This was written for DomPDF 1.0.2 and it worked very well. 

The rationale behind this hook is to automatically calculate the page top and bottom margin based on header and footer height and set them. Whenever a header or footer element is found on page 0, it is "reflown()". This does some kind of positioning of the element and its children, and then the height of all positioned Block elements inside are added and stored ($headerheight, $footerheight). The stored values are then added/subtracted from the page drawable area by setting them as margins. This way you don't have to set header or footer heights to any value - they are just auto. Same goes for the fake page margins. 

On the fly we count all pages created. The second hook checks replaces variables {PG} & {PC} to the number of pages or the current page.

Maybe someone can port this over to DomPDF 2.0.4?

```php
$dompdf->setCallbacks([
            [
                'event' => 'begin_page_reflow',
                'f' => function($params) use (&$pagecount, &$headerheight, &$footerheight) {
                    $frame = $params['frame'];
                    if ($pagecount == 0) {
                        foreach ($frame->get_children() as $child) {
                            if (!in_array($child->get_node()->nodeName, ['header', 'footer'])) continue;
                            if ($child->get_style()->position != 'fixed') continue;
                            $child->reflow();
                            $block = $child->get_frame()->get_decorator();
                            if (!$block instanceof \Dompdf\FrameDecorator\Block) continue;
                            foreach ($block->get_line_boxes() as $line) {
                                if ($child->get_node()->nodeName == 'header') $headerheight += $line->h;
                                if ($child->get_node()->nodeName == 'footer') $footerheight += $line->h;
                            }
                            $child->reset();
                        }
                    }
                    $box = $frame->get_containing_block();
                    $box['y'] += $headerheight;
                    $box['h'] -= $headerheight;
                    if ($pagecount == 0) $box['h'] -= $footerheight;
                    $frame->set_containing_block($box['x'], $box['y'], $box['w'], $box['h']);
                    $root = $frame->get_root();
                    $root->get_style()->margin_bottom = ((float)$root->get_original_style()->margin_bottom + ($footerheight * 1.33333)) . 'px';
                    $pagecount++;
                },
            ],
            [
                'event' => 'my_variables',
                'f' => function($params) {
                    $canvas = $params['canvas'];
					$page = 0;
                    foreach ($canvas->get_cpdf()->objects as &$object) {
                        if ($object['t'] == 'page') $page++;
                        if ($object['t'] == 'contents') $object['c'] = str_replace([static::PAGECOUNT, static::PAGE], [$canvas->get_page_count(), $page], $object['c']);
                    }
                },
            ]
        ]);
```

**Reply by bsweeney** [reply]
Are you asking for help porting to 2.x? The primary thing you need to change is the callable's signature from `function($params)` to `function ($frame, $canvas, $fontMetrics)` and then update the usage within the methods.

The second event (my_variables) looks to be a customization to the source since it's not standard for Dompdf.

**Reply by procodix** [reply]
tried it. Works. However it seems that something changed in DomPDF 2. A fixed header with top:0px is not positioned absolutely on the page but relatively to the body.

**Reply by bsweeney** [reply]
Positioned elements have always been relative to the body, not the page itself. If you want to push the header outside the bounds of the body use a negative top/bottom.

**Comment by procodix** [comment]
It fails for multiple pages. Don't know how the header & footer elements are copied to every page, but the hook doesn't see them at all, so I can't alter the positions.

**Comment by procodix** [comment]
header is blue, footer is green, body is red.

Page 1 looks good, besides some minor overlapping of body and footer:
<img width="580" alt="image" src="https://github.com/dompdf/dompdf/assets/17002625/e9c79799-c22e-4088-b4ef-d2cb85870afc">

Page 2 doesn't set the header top and footer bottom positioning, body seems to high:
<img width="577" alt="image" src="https://github.com/dompdf/dompdf/assets/17002625/00b53a04-4987-4b95-bdcc-49d6714ce432">

Page 3 does something even stranger:
<img width="576" alt="image" src="https://github.com/dompdf/dompdf/assets/17002625/c310ce92-4c91-48c7-befe-ed5240d4d936">

**Reply by bsweeney** [reply]
I would need to see sample of your HTML + CSS and related PHP code to be able to hazard a guess as to what's going on.

**Reply by procodix** [reply]
Sorry, cant provide full html, but it is a very simple structure:
html
  body
     header (fixed top)
     footer (fixed bottom)
     content (paragraphs and a table that spans multiple pages)

**Reply by bsweeney** [reply]
no I get that but without seeing the actual structure and HTML used I can't really provide any guidance. It doesn't have to be an actual copy of your document, just something that represents what you're trying to do.
